`instr` declaration parameter mapping (#1044)
powdr-labs/powdr: book/src/asm/instructions.md

```
machine Main {
    ...
    reg X[<=];
    reg Y[<=];
    reg Z[<=];

    instr add X, Y -> Z = submachine.add; // will call submachine.add(X, Y) and return through Z
    // this is equivalent to:
    // instr add X, Y -> Z = submachine.add X, Y -> Z;
    ...
}
```

The mapping is flexible and allows using, besides the assignment registers in the instruction parameters, any regular registers from the local machine (i.e., the caller).
All the following mappings are possible:
```
machine Main {
    reg X[<=];
    reg Y[<=];
    reg Z[<=];
    reg A;
    reg B;
    reg C;
    ...
    instr add_to_A X, Y = addvm.add X, Y -> A;          // - output written to a regular register
    instr addAB -> X = addvm.add A, B -> X;             // - input from the value of regular registers
    instr addAB_to_C = addvm.add A, B -> C;             // - regular registers used for both inputs and output
    instr addAB_to_A = addvm.add A, B -> A;             // - read from and then write to the same regular register
    instr sub_from_add X, Y -> Z = addvm.add Y, Z -> X; // - find input for a given output (doesn't always work!)
    ...
}

> ```

Reviewer: I think we should also mention somewhere that assignment registers have to be "repeated" on the lhs while non-assignment registers can only appear on the rhs. Not sure how to say that in an understandable way, though :)

# External instructions

An external instruction delegates its implementation to a function/operation from a submachine.
When called, the inputs and outputs of the declared instruction are mapped into a call to the submachine function.

Assume we have a submachine with a single operation `add`:
```
machine SubMachine(latch, operation_id) {
    operation add<0> a,b -> c;
    ...
}
```

An external instruction calling into this operation can be declared as follows:
```
machine Main {
    ...
    reg X[<=];
    reg Y[<=];
    reg Z[<=];

    instr add X, Y -> Z = submachine.add; // will call submachine.add(X, Y) and return through Z
    // this is equivalent to:
    // instr add X, Y -> Z = submachine.add X, Y -> Z;
    ...
}
```

> The left-hand side of the definition declares the instruction parameters, that is, how the instruction is called in the local machine.

Reviewer: ```suggestion
The left-hand side of the definition states which local registers to use for calling the instruction.
```
(in my understanding, "declare" introduces a new symbol)

> The right-hand side of the definition declares the external operation being called, and how the instruction parameters are used in that call.

Reviewer: ```suggestion
The right-hand side of the definition specifies  the external operation being called, and how the instruction parameters are used in that call.
```

Reviewer: Hm, I fear this still does not really catch the essence...
But I also don't have a better suggestion right now.

Reviewer: Maybe something along the lines of "left hand side defines the name of the local instruction and and its parameters, who all need to be assignment registers. The right hand side specifies the name of the sub-machine instruction and how to call it. Here, in addition to the assignment registers used on the left hand side, non-assignment registers can be used. Those can be seen as implicit parameters."

Author: reworked the text and expanded the example

In the above example, parameters of the instruction match one to one, in order, with the call to the target operation, and can thus be omitted.
Otherwise, every parameter in the instruction (left-hand side) must be used in the external call (right-hand side).
The mapping is flexible though, and allows parameters to be reordered and the use of additional regular registers when calling the external operation.
All the following mappings are possible:
```
machine Main {
    reg X[<=];
    reg Y[<=];
    reg Z[<=];
    reg A;
    reg B;
    reg C;
    ...
    instr add_to_A X, Y = addvm.add X, Y -> A;          // - output written to a regular register
    instr addAB -> X = addvm.add A, B -> X;             // - input from the value of regular registers
    instr addAB_to_C = addvm.add A, B -> C;             // - regular registers used for both inputs and output
    instr addAB_to_A = addvm.add A, B -> A;             // - read from and then write to the same regular register
    instr sub_from_add X, Y -> Z = addvm.add Y, Z -> X; // - find input for a given output (doesn't always work!)

>     ...

Reviewer: Maybe also add examples how the instruction are used.
